extmod/btstack/modbluetooth_btstack.c

```diff
@@ -1109,90 +1069,152 @@ int mp_bluetooth_gatts_write(uint16_t value_handle, const uint8_t *value, size_t
-    return mp_bluetooth_gatts_notify_send(conn_handle, value_handle, data, len);
+    return 0;
 }
 
-int mp_bluetooth_gatts_notify_send(uint16_t conn_handle, uint16_t value_handle, const uint8_t *value, size_t value_len) {
-    DEBUG_printf("mp_bluetooth_gatts_notify_send\n");
+#if !MICROPY_TRACKED_ALLOC
+#error "btstack requires MICROPY_TRACKED_ALLOC"
+#endif
 
-    if (!mp_bluetooth_is_active()) {
-        return ERRNO_BLUETOOTH_NOT_ACTIVE;
-    }
+typedef struct {
+    btstack_context_callback_registration_t btstack_registration;
+    int gatts_op;
+    uint16_t conn_handle;
+    uint16_t value_handle;
+    size_t value_len;
+    uint8_t value[];
+} notify_indicate_pending_op_t;
 
-    // Attempt to send immediately. If it succeeds, btstack will copy the buffer.
+// Called in response to a gatts_notify/indicate being unable to complete, which then calls
+// att_server_request_to_send_notification.
+STATIC void btstack_notify_indicate_ready_handler(void *context) {
     MICROPY_PY_BLUETOOTH_ENTER
-    int err = att_server_notify(conn_handle, value_handle, value, value_len);
-    MICROPY_PY_BLUETOOTH_EXIT
-
-    if (err == BTSTACK_ACL_BUFFERS_FULL) {
-        DEBUG_printf("mp_bluetooth_gatts_notify_send: ACL buffer full, scheduling callback\n");
-        // Schedule callback, making a copy of the buffer.
-        mp_btstack_pending_op_t *pending_op = btstack_enqueue_pending_operation(MP_BLUETOOTH_BTSTACK_PENDING_NOTIFY, conn_handle, value_handle, value, value_len);
-
-        err = att_server_request_to_send_notification(&pending_op->context_registration, conn_handle);
-
-        if (err != ERROR_CODE_SUCCESS) {
-            // Failure. Unref and free the pending operation.
-            btstack_remove_pending_operation(pending_op, true /* del */);
-        }
-
-        return 0;
-    } else {
-        return btstack_error_to_errno(err);
+    notify_indicate_pending_op_t *pending_op = (notify_indicate_pending_op_t *)context;
+    DEBUG_printf("btstack_notify_indicate_ready_handler gatts_op=%d conn_handle=%d value_handle=%d len=%lu\n", pending_op->gatts_op, pending_op->conn_handle, pending_op->value_handle, pending_op->value_len);
+    int err = ERROR_CODE_SUCCESS;
+    switch (pending_op->gatts_op) {
+        case MP_BLUETOOTH_GATTS_OP_NOTIFY:
+            err = att_server_notify(pending_op->conn_handle, pending_op->value_handle, pending_op->value, pending_op->value_len);
+            DEBUG_printf("btstack_notify_indicate_ready_handler: sending notification err=%d\n", err);
+            break;
+        case MP_BLUETOOTH_GATTS_OP_INDICATE:
+            err = att_server_indicate(pending_op->conn_handle, pending_op->value_handle, pending_op->value, pending_op->value_len);
+            DEBUG_printf("btstack_notify_indicate_ready_handler: sending indication err=%d\n", err);
+            break;
     }
+    assert(err == ERROR_CODE_SUCCESS);
+    (void)err;
+    m_tracked_free(pending_op);
```

Comment:
Should this free go outside `MICROPY_PY_BLUETOOTH_EXIT?

Reply:
Yes. Fixed.